src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py

```diff
@@ -14,6 +16,125 @@
+            if msg.dataset in dataset_config.engagement_db_datasets:
+                dataset_messages.append(msg)
+
+        # If there are no messages in this dataset, either clear the contact field if we're allowed to, or simply skip
+        # this dataset if not.
+        # (We might not be able to be allowed to clear the field because doing so could cause synchronisation problems
+        #  with Rapid Pro, due to the delay between a flow updating a contact field and that change being processed
+        #  by our infrastructure and synced back - in other words, we can't guarantee read-after-write consistency
+        #  between Rapid Pro and the engagement database)
+        if len(dataset_messages) == 0:
+            if sync_config.allow_clearing_fields:
+                contact_fields[dataset_config.rapid_pro_contact_field.key] = ""
+            continue
+
+        # Write the found data back to Rapid Pro as either, depending on the configuration, the raw data
+        # (better for debugging) or as presence indicators (stronger privacy preservation).
+        if sync_config.write_mode == WriteModes.SHOW_PRESENCE:
+            contact_fields[dataset_config.rapid_pro_contact_field.key] = _PRESENCE_VALUE
+        else:
+            assert sync_config.write_mode == WriteModes.CONCATENATE_TEXTS
+            message_strings = [f"\"{msg.text}\" - engagement_db.{msg.dataset}" for msg in dataset_messages]
+            contact_fields[dataset_config.rapid_pro_contact_field.key] = "; ".join(message_strings)
+
+    return contact_fields
+
+
+def _get_consent_withdrawn_field_for_participant(participant_messages, sync_config, code_schemes):
+    """
+    Gets the consent_withdrawn contact field for a given participant and sync configuration.
+
+    :param participant_messages: All messages from the participant to process.
+    :type participant_messages: list of engagement_database.data_models.Message
+    :param sync_config: Sync config defining which messages to get.
+    :type sync_config: src.engagement_db_to_rapid_pro.configuration.EngagementDBToRapidProConfiguration
+    :param code_schemes: Project code schemes (used to decode the labels to identify consent withdrawn messages).
+    :type code_schemes: list of core_data_modules.data_models.CodeScheme
+    :return: Dictionary of Rapid Pro contact field id -> value.
+    :rtype: dict of str -> str
+    """
+    if sync_config.consent_withdrawn_dataset is None:
+        return dict()
+
+    all_labels = []
+    for msg in participant_messages:
+        if msg.dataset in sync_config.consent_withdrawn_dataset.engagement_db_datasets:
+            all_labels.extend(msg.get_latest_labels())
+
+    contact_fields = dict()
+    consent_withdrawn_contact_field = sync_config.consent_withdrawn_dataset.rapid_pro_contact_field
+    if _labels_contain_consent_withdrawn(all_labels, code_schemes):
+        contact_fields[consent_withdrawn_contact_field.key] = "yes"
+    elif sync_config.allow_clearing_fields:
```

Comment:
Which use case would we want to clear the consent_withdrawn_contact_field?

Reply:
In cases where it had previously been set incorrectly e.g. if someone labels a message as STOP then corrects it to something else.

Reply:
Thanks makes sense.